### PR TITLE
Issue#1936 annotation

### PR DESCRIPTION
**Pull request by spring-ryanc**
Made lots of changes! It now assigns the glow class to the spans and splits them if needed. Made changes to method calls and added a migration file to support inclusion of column offsets. Also fixed an issue where the mouse listeners weren't removed if the line was no longer glowing

The glow method is pretty long right now but I didn't really see many opportunities to break it into more methods, open to suggestions.

Also I assume there will be some formatting issues, hopefully hound/travis let me know so I can make the changes ASAP.

### CODE REVIEW

**Comment by david-yz-liu** [comment]
I think the tests need to be updated. Changes look cool - I'll take a closer look later this week.

File: app/assets/javascripts/SourceCodeGlower/SourceCodeLine.js [hunk 1]
```diff
@@ -22,26 +24,201 @@ function SourceCodeLine(line_node) {
...
+    var endCharCount = currentCharCount + textNodes[i].length;
+
+    // If start is between current count and end of the current node, save it
+    if (!foundStart && start >= currentCharCount && start <= endCharCount) {
+      foundStart = true;
+
+      // Will need to split the node if the glow starts inside of it
+      if (start != endCharCount) {
+        // Save reference to span node, create span wrapper if blank space
+        if ( textNodes[i].parentNode.parentNode == node){
```
Comment (david-yz-liu) [inline app/assets/javascripts/SourceCodeGlower/SourceCodeLine.js]: The whitespace here is odd - no space after `(`, space before `{`.

Also, you should use `===` here

File: app/assets/javascripts/SourceCodeGlower/SourceCodeLine.js [hunk 2]
```diff
@@ -22,26 +24,201 @@ function SourceCodeLine(line_node) {
...
+          textNodes[i] = start_node.childNodes[0];
+        }
+        else {
+          start_node = textNodes[i].parentNode;
+        }
+        start_node_offset = start - currentCharCount;
+      }
+    }
+    // Adds class/data/events to the nodes between the start and end
+    else if (foundStart && end >= endCharCount){
```
Comment (david-yz-liu) [inline app/assets/javascripts/SourceCodeGlower/SourceCodeLine.js]: Space before `{`

File: app/assets/javascripts/SourceCodeGlower/SourceCodeLine.js [hunk 3]
```diff
@@ -22,26 +24,201 @@ function SourceCodeLine(line_node) {
...
+      var glow_depth = textNodes[i].parentNode.getAttribute("data-annotationDepth");
+      textNodes[i].parentNode.setAttribute("data-annotationDepth",
+        glow_depth == null ? "1" : (parseInt(glow_depth) + 1).toString());
```
Comment (david-yz-liu) [inline app/assets/javascripts/SourceCodeGlower/SourceCodeLine.js]: Specify a radix of 10 when you call `parseInt`: `parseInt(glow_depth, 10)`.

File: app/views/results/common/_text_codeviewer.html.erb [hunk 4]
```diff
@@ -26,7 +26,7 @@
         add_annotation_text(<%= annot.annotation_text.id %>,
                             '<%= simple_format(escape_javascript(annot.annotation_text.content.to_s)) %>');
         add_annotation(<%= annot.id %>,
-                       { start: <%= annot.line_start %>, end: <%= annot.line_end %> },
+                       { start: <%= annot.line_start %>, end: <%= annot.line_end %>, column_start: <%= annot.column_start %>, column_end: <%= annot.column_end %> },
```
Comment (david-yz-liu) [inline app/views/results/common/_text_codeviewer.html.erb]: This line is too long.

**Comment by david-yz-liu** [comment]
@spring-ryanc This is looking quite impressive. My main thought on splitting up `glow` is to break up each conditional block into its own method. I don't think many parameters would need to be passed into each helper, though I could be wrong.

**Comment by spring-ryanc** [comment]
Hi David, I've made the style changes you suggested. But I'm not sure which test need to be updated. Could you provide a little more information about that?

**Comment by david-yz-liu** [comment]
You can see from the Travis status: https://travis-ci.org/MarkUsProject/Markus/builds/49055684

**Comment by david-yz-liu** [comment]
@spring-ryanc Looks good, although I think Hound would like you to change the formatting of that blueprints.rb file. :)

**Comment by david-yz-liu** [comment]
@spring-ryanc I merged it, but looks like multi-line annotations aren't working now. Also, I tried double-clicking a word and adding an annotation, and the annotation was created, but added to the wrong location.

**Comment by spring-ryanc** [comment]
Odd, those both worked locally, must be an issue with my git/commit. I'll figure out when I'm back home in a couple days. Thanks

**Pull request merged**